Fix reuse of `ConsolidateBlocks` instances

jakelishman opened the pull request:
The `_qubit_map` state added to `ConsolidateBlocks` in #15083 (cb103d95) was attempting to get round an awkward internal `PassManager` use to pass the mapping into recursive calls.  Unfortunately, the state was not cleared on exit to the pass, so re-use of the pass object would be invalid, such as by calling the `PassManager.run` method on the result of `generate_preset_pass_manager` more than once on circuits with different sizes.

Rather than trying to plaster over the poor use of state, this commit instead moves the runtime logic to be in the `PropertySet`, so re-use troubles are limited to single executions.  It's still technically possible for the `PropertySet` to become polluted if `ConsolidateBlocks` throws an internal exception that is forcibly suppressed such that the pipeline continues, but this is completely non-standard, would require significant effort to achieve, and most likely there are large numbers of transpiler passes that are not completely exception-safe anyway.

<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary



### Details and comments

Fix #15255

jakelishman committed:
Fix reuse of `ConsolidateBlocks` instances

The `_qubit_map` state added to `ConsolidateBlocks` in #15083 (cb103d95)
was attempting to get round an awkward internal `PassManager` use to
pass the mapping into recursive calls.  Unfortunately, the state was not
cleared on exit to the pass, so re-use of the pass object would be
invalid, such as by calling the `PassManager.run` method on the result
of `generate_preset_pass_manager` more than once on circuits with
different sizes.

Rather than trying to plaster over the poor use of state, this commit
instead moves the runtime logic to be in the `PropertySet`, so re-use
troubles are limited to single executions.  It's still technically
possible for the `PropertySet` to become polluted if `ConsolidateBlocks`
throws an internal exception that is forcibly suppressed such that the
pipeline continues, but this is completely non-standard, would require
significant effort to achieve, and most likely there are large numbers
of transpiler passes that are not completely exception-safe anyway.

mtreinish commented on qiskit/transpiler/passes/optimization/consolidate_blocks.py:
```diff
@@ -139,9 +138,9 @@ def run(self, dag):
-        if self._qubit_map is None:
-            self._qubit_map = list(range(dag.num_qubits()))
-
+        qubit_map = self.property_set.get(self._QUBIT_MAP_KEY, None)
```
I prefer being explicit like this but isn't `property_set` default dict like and return `None` on getitem for a missing entry anyway?

jakelishman replied:
It is, but relying on that has the side-effect of creating the key if it didn't already exist, which I was vaguely trying to avoid given the statefulness problems here.

mtreinish commented on test/python/compiler/test_transpiler.py:
```diff
@@ -2333,6 +2333,53 @@ def test_single_qubit_circuit_deterministic_output(self, optimization_level):
+
+        circuits = [ghz(5), ghz(8), ghz(10)]
+        shared_pm = make_pm()
+        shared_circuits = [shared_pm.run(circuit) for circuit in circuits]
+        own_circuits = [make_pm().run(circuit) for circuit in circuits]
+        self.assertEqual(shared_circuits, own_circuits)
```
Do you want to add a variant of this which passes a list to a single pass manager, like: `shared_circuits = shared_pm.run(circuits)`? I would have assume we have coverage of this somewhere else, but considering this bug slipped through maybe we only test it in a multiprocessing context?

jakelishman replied:
Sure, doesn't hurt.  Done in bb4f7e81c8.

mtreinish commented on qiskit/transpiler/passes/optimization/consolidate_blocks.py:
```diff
@@ -68,6 +68,8 @@ class ConsolidateBlocks(TransformationPass):
         collected by a previous pass, such as `Collect2qBlocks`.
     """
 
+    _QUBIT_MAP_KEY = "ConsolidateBlocks_qubit_map"
```
While I know this is private and only used internally for the control flow path, do you want to document that the pass reads this? Just in case someone tried to use this a conflicting key?

jakelishman replied:
Oh yeah thanks, I probably should have.  Done in bb40bed92f.

jakelishman committed:
Add test of `PassManager.run` over a list

On some platforms this will be multiprocessing, but it doesn't hurt to
try it for the platforms that aren't.

jakelishman committed:
Document use of property-set key

mtreinish reviewed (approved):
LGTM now, thanks for the quick updates
